### ISSUE TITLE
Suggestion

### ISSUE THREAD

**Issue by boriborm** [pljson/pljson#148]
I have suggestion

add in `pljson` type function `put` who return `self` as `pljson`

Then we can write code as:

```
declare
 j pljson;
begin
 j:=pljson()
      .put('a','a')
      .put('b',pljson()
               .put('c','c')
               .put('d','d')
      );
dbms_output.put_line(j.to_char());
end;

// result:   { a : "a", b : { c : "c" , d : "d" } }

```
but it is not good idea for big values (clobs). Each `put` function create new json instance object.

**Comment by jsumners** [comment]
Go for it. At this point, I consider you and @dsnz the real maintainers of this project. I no longer use an Oracle database, so I can't even run tests.

**Comment by dsnz** [comment]
@boriborm 
it can be done very easily and yes it is not efficient and that's why I did not implement it
instead I promote use of new constructors using arrays, like
```constructor function pljson(str_array pljson_varray) return self as result```

but I agree it's fancy (builder style) coding and should be an option for developers who know what they are doing

go ahead and provide an implementation if you like and you feel it's worthy

**Comment by dsnz** [comment]
@boriborm 
it can be done very easily it can be efficient if done carefully
but I like use of new constructors using arrays, like
```constructor function pljson(str_array pljson_varray) return self as result```

but I agree it's fancy (builder style) coding and should be an option for developers who know what they are doing

go ahead and provide an implementation if you like and you feel it's worthy

**Comment by boriborm** [comment]
pljson_array can work only with varchar2.

In my case we can create pljson with any data types.
```
declare
 j pljson;
begin
 j:=pljson()
      .p('a',123)
      .p('b',pljson()
               .p('c',true)
               .p('d',123)
      )
     .p('list', pljson_list()
         .a(123)
         .a('test')
         .a(pljson()
                  .p('z','z')
         )
    )
;
dbms_output.put_line(j.to_char());
end;

// result:   { a : 123, b : { c : true , d : 123 } , list : [ 123 , "test" , { z : "z"}  ] }
```

**Comment by dsnz** [comment]
yes indeed
(the array constructor handles the most common case since varrays must be of same type,
another constructor could handle arrays of numbers etc.)
I believe the savings in characters with chaining is not so important
but there are other users who have asked the same thing so perhaps this construction chaining should be added

please go ahead and provide an implementation

**Comment by dsnz** [comment]
hello,
great but I have a few suggestions before merging

1. I think there is no need for a new type, you can new "g" and "p" member methods to the pljson object
(a new type would be unnecessary confusion to users who already must grasp:
```pljson```, ```pljson_list```, ```pljson_value```)
2. you need not call the constructor again in the putters,
you can return the same object
see https://stackoverflow.com/questions/18086603/how-to-chain-calls-in-a-pl-sql-object-type-of-functions-returning-self
and the following trivial demo implementation for ideas
```
create or replace type t force as object (

  a number,

  constructor function t return self as result,
  member function inc return t
)

create or replace type body t as

constructor function t return self as result is
begin
  a := 20;
  return;
end;

member function inc return t IS
  v_t t := self;
begin
  v_t.a := v_t.a + 1;
  return v_t;
end;

end;

declare
  x_t t;
begin
  x_t := t().inc().inc().inc();
  dbms_output.put_line(x_t.a);
end;
-- this returns 23, 20 incremented 3 times
```

**Comment by boriborm** [comment]
I thought it is a specific coding style, so I did it as an addon. `pljson` is a base type and the more functions and procedures, the more need memory for object creation (may be I wrong ?).
For getters (`g` functions) need fake procedure (`getter`) where getters return last value.

**Comment by dsnz** [comment]
don't worry about memory, extra methods do not use extra memory
(after all, the new pljson_factory object encapsulates a pljson object so needs a tad bit more memory)

as I see, you don't need the getters, they just return the internal pljson object or other type manifestations of it, something the developer can do this himself/herself and does not relate with the proposed coding style

so you need only implement the putters (```p()```) or adders (```a()```, for pljson_list, see below)
also you must do this for pljson_list too (something that is not handled well with the _factory object)
see your proposed example which has a ```pljson_list().a(123)```

so to summarize, just implement methods

**p() for pljson and a() for pljson_list**

and your example will work instantly without change and the developers will be able to work more naturally and easily

**Comment by boriborm** [comment]
I need getters.
It's fastest write code than use pljson_value or pljson_ext.
For example:

```
declare
 j pljson:=pljson('{
    a:"a",
    b:123,
    c:{c:"c"},
    d:"01.01.2018"
 }');
 a varchar2(100);
 b number;
 c varchar2(100);
 d date;
 j2 pljson; 
begin
    /* use factory */
    pljson_factory.getter(
        pljson_factory(j)
            .g('a',a)        
            .g('b',b)
            .g('c', j2)
            .g('d',d,'dd.mm.rrrr')
            .get_json('c')
                .g('c', c)
     );
     
     /* use pljson_value */
     a:=j.get('a').get_string();
     b:=j.get('b').get_number();
     j2:=pljson(j.get(c));
     c:=j2.get('c').get_string();
     d:=to_date(j.get('d').get_string(),'dd.mm.rrrr');
     
     /* use pljson_ext */
     a:=pljson_ext.get_string(j,'a');
     b:=pljson_ext.get_number(j, 'b');
     j2:=pljson_ext.get_json(j,'c');
     c:=pljson_ext.get_string(j,'c.c');
     d:=to_date(pljson_ext.get_string(j,'d'),'dd.mm.rrrr');
end;
```

getters can be in pljson type, but it need fake procedure (or use pljson variable for returning value) and variable for parent pljson for `up` function

**Comment by dsnz** [comment]
I understand very well what the code marked with

```/* use pljson_value */```

will return
it works already existing code already and will work nicely with new **p() and a()**

but I can't imagine what the proposed code marked with

```/* use factory */```

will return

can you please make more clear what will be the result of ```pljson_factory.getter(...);``` ?

**Comment by boriborm** [comment]
getters (`g` functions) read tag values and set variables.

getters is a functions

I can't write
```
 pljson_factory(j).g('a',a).g('b',b);
```
result of function must be set in variable or used in other function/procedure

to use getters without fake procedure we need make variable or use procedures. Example:
```
declare
 j pljson;
 a varchar2(100);
 b number;
 jf pljson_factory;
begin
 j:=pljson_factory(pljson('{a:"a", b:123}')).g('a',a).g('b',b).get(); -- worked. Need use pljson type variable (j)
 jf:=pljson_factory(pljson('{a:"a", b:123}')).g('a',a).g('b',b); -- worked. Need use pljson_factory variable (jf)

 dbms_output.put_line (pljson_factory(pljson('{a:"a", b:123}')).g('a',a).g('b',b).get().to_char()); --worked. Used procedure.

 pljson_factory(pljson('{a:"a", b:123}')).g('a',a).g('b',b); --not worked!!! g is a function!!!

 -- to get values without variables as j or jf we need use some procedure. But what?
 -- pljson_factory has getter procedure that do nothing
 pljson_factory.getter( 
        pljson_factory( pljson('{a:"a", b:123}')).g('a',a).g('b',b) 
 );

end;
```

**Comment by dsnz** [comment]
oh i think I got it !

the whole trick is in that the ```g()``` functions have a **in out** parameter to receive result
I didn't notice this first
the ```getter()``` function is just null and a trick to code a whole chain of g(s) ...

well this is very idiomatic but let's include it too 😄 

still you can implement the **g()** functions in ```pljson```
and even ```pljson_list``` too (in pljson_list as ```g(position pls_integer, value varchar2)``` etc.)

can you do such a PR ?

**Comment by boriborm** [comment]
I will try to do that :)

**Comment by dsnz** [comment]
thank you, it will be a great addition 👍

**Comment by dsnz** [comment]
@boriborm 

in the process of reviewing the code in version 3.0 (and also updating version 2.0 so that it has almost the same api calls available as version 3.0), I rethought the discussion here

I understand that you are concerned with extra and unnecessary object creations (though it is usually not much of a performance burden) and that's why you updated your original proposal so that a new **pljson_factory** object is used instead (and which contains a single pljson_object that is updated)
I agree now with this (which also helps in not putting many extra api calls in the main types that exist now as you noted)

so I hope you proceed and provide a PR for new _factory or _builder types (_builder is better I think because this is the proper pattern that is used here)

**pljson_factory** or **pljson_builder** with methods p() and g() (for varchar2, clob, number, binary_double, boolean, pljson, pljson_list)

e.g.
```member function p(self in out nocopy pljson_builder, pair_name varchar2, pair_value varchar2, position pls_integer default null) return pljson_builder,```
```member function g(self in out nocopy pljson_builder, pair_name varchar2, pair_value out varchar2) return pljson_builder,```

**pljson_list_factory** or **pljson_list_builder** with methods a() and g() (for varchar2, clob, number, binary_double, boolean, pljson, pljson_list)

e.g.
```member function a(self in out nocopy pljson_list_builder, elem varchar2, position pls_integer default null) return pljson_list_builder,```
```member function g(self in out nocopy pljson_list_builder, position pls_integer, elem out varchar2) return pljson_list_builder,```

(there should be a PR for master and a PR for develop_v3)
what do you think ?

**Comment by julgon2012** [comment]
Maravilloso trabajo hacen.  Los felicito y los admiro!!!
